### ISSUE TITLE
RC5 Caching seems to *occasionally* return the wrong ID in Payload when cache_classes is true

### ISSUE THREAD

**Issue by hhff** [rails-api/active_model_serializers#1703]
Hi Team!~

I love AMS - thankyou everyone for the hard work!

This is more a PSA / Flag than an issue!  

TLDR: With `config.cache_classes = true` some very weird stuff happens with the AMS cache.

### **System**

active_model_serializers v0.10.0.rc5 @ JSONAPI & `:underscore` key transforms
ruby 2.2.1p85 (2015-02-26 revision 49769)
rails 4.2.2
puma
redis-rails

I have some weird behavior that I can't reproduce - I thought I'd post it here in case anyone has a lead on it.

### **How I introduced the bug**

- Update from `rc4` -> `rc5`
- Set `config.action_controller.perform_caching = true` & `config.cache_store = :redis_store` on the application level
- Introduce some light AMS caching + make necessary changes to configs
- Test everything on `development` & `staging` and all is good
- Deploy to Production
- Bug Appears

### **Buggy Behavior**

For the `Person` object:
```
{
  id: 13493,
  first_name: 'Hugh',
  last_name: 'Francis'
}
```

And the request:
`api/persons/13493`

**Very Rarely (maybe 1 in 30 requests or so)** the response will look like:

```
{
  "data": {
    "id": 84932,
    "type": "persons",
    "attributes": {
      "first_name": "Hugh",
      "last_name: "Francis"
    }
  }
}
```

**That's right, the response payload has the wrong ID, but the correct attributes (& relationships).**```

Important points:

- Bug is not related to **Thread Safety**.  The bug still occurs when running `Webbrick` in production (yup, I did that for a couple minutes)

### **How I Fixed It**

```config.cache_classes = false```

I'm guessing this has to do with the application reusing CachedSerializer Classes in different requests... but I dunno.

Figured I should flag this - as it took me 2 days to track down!

**Comment by beauby** [comment]
Thanks a lot for this @hhff – will definitely be helpful.

**Comment by bf4** [comment]
This is great! If you can provide even a backtrace that would be helpful.. without a test case it's harder to reproduce this and fix it

**Comment by hhff** [comment]
Hmmm - that could be difficult.  I might be able to provide a reproduction repo tho!

**Comment by bf4** [comment]
let us know :)

**Comment by hhff** [comment]
Will look into it later this week / early next!

**Comment by onomated** [comment]
Thanks @hhff 
I'm experiencing this also.  Same setup. `redis-cache`, `puma`, `rails 4.2.2`.  I'm getting wrong attributes associated with ids, and cache excluded fields seem to reflecting invalid cached state.  So:
```
cache({skip_digest: true, except: [:num_followers]})
```

`num_followers` isn't getting updated even though I've confirmed that the corresponding redis field is updated and propagated to the point of rendering the response.

Says this is `Ready for PR` is there actually a fix?  Or just documentation to disable `cache_classes` config?

**Comment by hhff** [comment]
phew - I thought I was losing my mind.  Just spent today building a stress test for this - but haven't been able to reproduce it yet.

@onomated - can you confirm that disabling `cache_classes` fixed the issue?

**Comment by onomated** [comment]
@hhff So far it does with the id -> attributes mismatch.  My cache exclude issues may be related to the fact that I override the `object_cache_key` method in my serializers.  Some of my models are cached under the expected keys and some are under the original AMS defaults.  I'll continue testing to see if these are isolated issues

**Comment by onomated** [comment]
@hhff from what I can tell disabling `cache_classes` fixes the issue with the id mismatch, but I haven't been able to reproduce empirically, just performed multiple deploys to production and the issue hasn't popped up.  Just worried about the effect on the rails app as a whole. 

Other "issue" I mentioned with `except` option seems to be a completely different problem related to fragment caching.  `object_cache_key` is not respected with fragment caching i.e. cached responses use AMS default keys vs my custom representation which results in overwritten cache results.

**Comment by hhff** [comment]
Yeah - I mean it would be nice to be able to use `cache_classes` with AMS.

So far I haven't had any luck reproducing the bug with a vanilla rails app.  I'm guessing that you need a tangled bulk of models and requests to trip it up...

**Comment by onomated** [comment]
From the sounds of it in this [blog post](http://tenderlovemaking.com/2012/06/18/removing-config-threadsafe.html), `cache_classes` should contribute to thread safety especially if rails is eager loaded on boot.  Haven't encountered this since, but definitely has look like a race condition somewhere

**Comment by onomated** [comment]
@hhff, do you happen to have fragment caching in any of your serializers i.e. specify (except or only cache options).  Prior implementations dynamically created classes to handle this.  This may have something to do with thread safety.  This has since been changed in master branch

**Comment by hhff** [comment]
@onomated - I ran my server with Webbrick (single threaded) and still got these errors, so it doesn't seem related to threadsafety - but most likely leaky memory allocations.

I did notice this commit: https://github.com/rails-api/active_model_serializers/pull/1622

My guess is that's our culprit

**Comment by hhff** [comment]
And yes I do use fragment caching

**Comment by bf4** [comment]
Have any benchmarks that show perf improvement w fragment caching on?  I've been working on it a lot but still is slower that not caching or full caching 

Original benchmark when introduced isn't available 

B mobile phone

**Comment by onomated** [comment]
@hhff ah yes, that logic has been [changed](https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model/serializer/caching.rb#L233) 

@bf4 No benchmarks unfortunately.  The fields I exclude from caching are all stored in redis, same as my cache so I'm **hoping** it will almost be as fast as fully cached, and definitely faster than non-caching.  Don't have numbers for this though but its worth knowing if its best to just leave fragmented entities entirely non-cached.

**Comment by hhff** [comment]
@bf4 I don't have any useful benchmarks - but can confirm that when I released our caching layer (which is heavily fragment cached in places) we noticed a big slowdown (2x in places) on our monitoring (skylight.io).  I've been hoping it would shake out, and been tweaking things here and there - but can confirm that it slowed the app.

**Comment by hhff** [comment]
@onomated - please let us know your findings!
